### PR TITLE
Rebased LZ4 PR

### PR DESCRIPTION
**Pull request by michael-grunder**
This is the LZ4 PR by @iliaal + some paranoid reworking by me.

@yatsukhnenko Will find all my bugs :smile:

Original PR #1754

### CODE REVIEW

File: common.h [hunk 1]
```diff
@@ -100,6 +102,7 @@ typedef enum {
 #define REDIS_COMPRESSION_NONE 0
 #define REDIS_COMPRESSION_LZF  1
 #define REDIS_COMPRESSION_ZSTD 2
+#define REDIS_COMPRESSION_LZ4 3
```
Comment (yatsukhnenko) [inline common.h]: Indentation

File: common.h [hunk 2]
```diff
@@ -274,6 +277,8 @@ typedef struct {
     redis_serializer   serializer;
     int                compression;
     int                compression_level;
+    int               compression_min_size;
```
Comment (yatsukhnenko) [inline common.h]: Indentation

File: library.c [hunk 3]
```diff
@@ -2280,6 +2302,29 @@ PHP_REDIS_API void redis_free_socket(RedisSock *redis_sock)
     efree(redis_sock);
 }
 
+#define PHP_REDIS_COMPRESSION_RATIO_CHECK(packed, unpacked) \
```
Comment (yatsukhnenko) [inline library.c]: I'm still not sure about compression minimums but I'm ok if you think that this is fine

File: library.c [hunk 4]
```diff
@@ -2397,6 +2510,41 @@ redis_unpack(RedisSock *redis_sock, const char *val, int val_len, zval *z_ret)
...
+
+                /* Read in our header bytes */
+                memcpy(&lz4crc, val, sizeof(uint8_t));
+                val += sizeof(uint8_t); val_len -= sizeof(uint8_t);
+                memcpy(&datalen, val, sizeof(int));
+                val += sizeof(int); val_len -= sizeof(int);
+
+                /* Make sure our CRC matches (TODO:  Maybe issue a docref error?) */
+                if (crc8((unsigned char*)&datalen, sizeof(datalen)) != lz4crc)
+                    break;
```
Comment (yatsukhnenko) [inline library.c:2523]: Shouldn't we try to unserialize data if decompression fails?

**Review by yatsukhnenko — APPROVED** [review]

File: library.c (hunk 3 above)
Comment (michael-grunder) [reply library.c]: How about I move it to a different PR and we can bikeshed it there :joy:

File: config.m4 [hunk 5]
```diff
@@ -194,6 +200,31 @@ if test "$PHP_REDIS" != "no"; then
...
+          break
+        fi
+      done
+      if test -z "$LIBLZ4_DIR"; then
+        AC_MSG_RESULT([not found])
+        AC_MSG_ERROR([Please reinstall the liblz4 distribution])
+      fi
+      PHP_CHECK_LIBRARY(lz4, LZ4_compress,
+      [
+        PHP_ADD_LIBRARY_WITH_PATH(zstd, $LIBLZ4_DIR/$PHP_LIBDIR, REDIS_SHARED_LIBADD)
```
Comment (orenwolf) [inline config.m4:219]: Should this mention `zstd` here?

Comment (michael-grunder) [reply config.m4:219]: No but this was fixed in 19269cb1db073376364d2655fc828daeb8f1006e

**Pull request merged**